### PR TITLE
[chore] [deltatocumulative]: linear histograms

### PR DESCRIPTION
**Pull request by tombrk**
#### Description

Finishes work started in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35048

That PR only partially introduced a less complex processor architecture by only using it for Sums.
Back then I was not sure of the best way to do it for multiple datatypes, as generics seemed to introduce a lot of complexity regardless of usage. 

I since then did of a lot of perf analysis and due to the way Go works (see gcshapes), we do not really gain anything at runtime from using generics, given method calls are still dynamic.

This implementation uses regular Go interfaces and a good old type switch in the hot path (ConsumeMetrics), which lowers mental complexity quite a lot imo. 

The value of the new architecture is backed up by the following benchmark:

```
goos: linux
goarch: arm64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor
                 │ sums.nested │             sums.linear             │
                 │   sec/op    │   sec/op     vs base                │
Processor/sums-8   56.35µ ± 1%   39.99µ ± 1%  -29.04% (p=0.000 n=10)

                 │  sums.nested  │             sums.linear              │
                 │     B/op      │     B/op      vs base                │
Processor/sums-8   11.520Ki ± 0%   3.683Ki ± 0%  -68.03% (p=0.000 n=10)

                 │ sums.nested │            sums.linear             │
                 │  allocs/op  │ allocs/op   vs base                │
Processor/sums-8    365.0 ± 0%   260.0 ± 0%  -28.77% (p=0.000 n=10)

```

#### Testing

This is a refactor, existing tests pass unaltered.

#### Documentation

not needed

### CODE REVIEW

File: processor/deltatocumulativeprocessor/internal/delta/delta.go [hunk 1]
```diff
@@ -96,6 +104,16 @@ func AccumulateInto[P data.Point[P]](state P, dp P) error {
...
+	case pmetric.NumberDataPoint:
+		state := any(state).(pmetric.NumberDataPoint)
+		data.Number{NumberDataPoint: state}.Add(data.Number{NumberDataPoint: dp})
+	case pmetric.HistogramDataPoint:
+		state := any(state).(pmetric.HistogramDataPoint)
+		data.Histogram{HistogramDataPoint: state}.Add(data.Histogram{HistogramDataPoint: dp})
+	case pmetric.ExponentialHistogramDataPoint:
+		state := any(state).(pmetric.ExponentialHistogramDataPoint)
+		data.ExpHistogram{DataPoint: state}.Add(data.ExpHistogram{DataPoint: dp})
+	}
```
Comment (sh0rez) [inline processor/deltatocumulativeprocessor/internal/delta/delta.go:117]: This refactor effectively eliminates the need for the data package, as we no longer rely on type characteristics.

I'll refactor datapoint addition in a future PR, making this part more clear, maybe like this:

```go
var add data.Aggregator = new(data.Add)

switch into := any(dp).(type) {
case pmetric.NumberDataPoint:
	add.Numbers(into, dp)
case pmetric.HistogramDataPoint:
	add.Histograms(into, dp)
}
```

**Review by RichieSams — APPROVED** [review]

File: processor/deltatocumulativeprocessor/benchmark_test.go [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+			require.NoError(b, err)
+		}
+	}
+
+	now := time.Now()
+	start := pcommon.NewTimestampFromTime(now)
+	ts := pcommon.NewTimestampFromTime(now.Add(time.Minute))
+
+	cases := []Case{{
+		name: "sums",
```
Comment (ArthurSens) [inline processor/deltatocumulativeprocessor/benchmark_test.go:72]: Is there a reason to split sums, histograms, and exponential histograms into separate benchmarks? Are those metric types expected to be split by separate deltatocumulative processors in real-world scenarios?

Comment (ArthurSens) [reply processor/deltatocumulativeprocessor/benchmark_test.go:72]: Your PR description only shows results for `sums`, so I'm not sure if this was an intentional split or you just forgot to benchmark the rest

File: processor/deltatocumulativeprocessor/benchmark_test.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+			}
+			req := pmetric.NewMetrics()
+			md.CopyTo(req)
+
+			b.StartTimer()
+			err := proc.ConsumeMetrics(ctx, req)
+			b.StopTimer()
+			require.NoError(b, err)
+		}
+	}
```
Comment (ArthurSens) [inline processor/deltatocumulativeprocessor/benchmark_test.go:65]: Is there any special reason to transform this into a function? We're not reusing the code anywhere, so why not just put this inside the b.Run loop?

File: processor/deltatocumulativeprocessor/benchmark_test.go [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+	const (
+		metrics = 5
+		streams = 10
+	)
+
+	type Case struct {
+		name string
+		fill func(m pmetric.Metric)
+		next func(m pmetric.Metric)
+	}
```
Comment (ArthurSens) [inline processor/deltatocumulativeprocessor/benchmark_test.go:37]: If we remove the abstraction of `run`, we could also move this closer to where it's used.

**Review by ArthurSens — COMMENTED** [review]
I did a first pass only through the benchmark. I totally understand my comments are nitpicks, I'm just sharing my personal preference when it comes to code style.

From our conversations I understand you prefer a more "declarative" style, but to me it makes the code way harder to read since I expect thing to run in the order they are written. Scrolling up and down several times until I finally understand what it does makes code less readable in my opinion.

Again, not a blocker!

File: processor/deltatocumulativeprocessor/internal/metrics/metrics.go [hunk 5]
```diff
@@ -63,3 +63,49 @@ func (m Metric) Typed() any {
...
 	panic("unreachable")
 }
+
+var (
+	_ Any = Sum{}
+	_ Any = Gauge{}
+	_ Any = ExpHistogram{}
+	_ Any = Histogram{}
+	_ Any = Summary{}
+)
```
[the remark this answers is not in the document]
Comment (sh0rez) [reply processor/deltatocumulativeprocessor/internal/metrics/metrics.go:73]: its a compile time assert those types implement `Any`

File: processor/deltatocumulativeprocessor/benchmark_test.go (hunk 2 above)
Comment (sh0rez) [reply processor/deltatocumulativeprocessor/benchmark_test.go:41]: it's interesting to see in the future if we make changes to the actual add implementations if certain datatypes get faster

File: processor/deltatocumulativeprocessor/benchmark_test.go [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+		gb.Run(cs.name, func(b *testing.B) {
+			st := setup(b, nil)
+			out = st.sink
+			run(b, st.proc, cs)
+
+			// verify all dps are processed without error
+			b.StopTimer()
+			if err := sdktest.Test(tel(b.N), st.tel.reader); err != nil {
+				b.Fatal(err)
+			}
```
[the remark this answers is not in the document]
Comment (sh0rez) [reply processor/deltatocumulativeprocessor/benchmark_test.go]: its a benchmark that checks it benchmarks the correct thing.
change to a simpler assert on output length

File: processor/deltatocumulativeprocessor/benchmark_test.go [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+	Ps interface {
+		At(int) P
+		Len() int
+	},
+	P interface {
+		Timestamp() pcommon.Timestamp
+		SetStartTimestamp(pcommon.Timestamp)
+		SetTimestamp(pcommon.Timestamp)
+	},
+](sel func(pmetric.Metric) T) func(m pmetric.Metric) {
```
[the remark this answers is not in the document]
Comment (sh0rez) [reply processor/deltatocumulativeprocessor/benchmark_test.go]: replaced by code duplication

**Comment by sh0rez** [comment]
@ArthurSens while the benchmark is split by datatype, I only ran for sums as the processor part (stream tracking, etc) is the same across all datatypes. speed differences only come by add implementation

**Review by ArthurSens — APPROVED** [review]

**Review by jpkrohling — APPROVED** [review]

**Pull request merged**